src/Common/src/Interop/Unix/System.Net.Security.Native/Interop.GssApi.cs

```diff
@@ -0,0 +1,136 @@
+            status = NetSecurity.InitSecContext(out minorStatus,
+                                                credential,
+                                                ref context,
+                                                isNtlm,
+                                                targetName,
+                                                (uint)inFlags,
+                                                buffer,
+                                                (buffer == null) ? 0 : buffer.Length,
+                                                out outputToken,
+                                                out outputLength,
+                                                out outFlags);
+
+            using (outputToken)
+            {
+                if ((status != NetSecurity.Status.GSS_S_COMPLETE) && (status != NetSecurity.Status.GSS_S_CONTINUE_NEEDED))
+                {
+                    throw new NetSecurity.GssApiException(SR.net_context_establishment_failed, status, minorStatus);
+                }
+
+                outputBuffer = new byte[outputLength];
+                if (outputLength > 0)
+                {
+                    Interop.NetSecurity.CopyBuffer(outputToken, outputBuffer, 0);
+                }
+            }
+            return (status == NetSecurity.Status.GSS_S_COMPLETE) ? true : false;
+        }
+
+        internal static int Encrypt(
+            SafeGssContextHandle context,
+            bool encrypt,
+            byte[] buffer,
+            int offset,
+            int count,
+            out byte[] outputBuffer)
+        {
+            outputBuffer = null;
+            Debug.Assert((buffer != null) && (buffer.Length > 0), "Invalid input buffer passed to Encrypt");
+            Debug.Assert((offset >= 0) && (offset < buffer.Length), "Invalid input offset passed to Encrypt");
+            Debug.Assert((count > 0) && (count <= (buffer.Length - offset)), "Invalid input count passed to Encrypt");
+
+            SafeGssBufferHandle outputToken;
+            int msgLength;
+            NetSecurity.Status minorStatus;
+            NetSecurity.Status status = NetSecurity.Wrap(out minorStatus, context, encrypt, buffer, offset, count, out outputToken, out msgLength);
+            using (outputToken)
+            {
+                if (status != NetSecurity.Status.GSS_S_COMPLETE)
+                {
+                    throw new NetSecurity.GssApiException(SR.net_context_wrap_failed, status, minorStatus);
+                }
+                outputBuffer = new byte[msgLength]; // Always return non-null
+                if (msgLength > 0)
+                {
+                    NetSecurity.CopyBuffer(outputToken, outputBuffer, 0);
+                }
+
+                return outputBuffer.Length;
+            }
+        }
+
+        internal static int Decrypt(
+            SafeGssContextHandle context,
+            byte[] buffer,
+            int offset,
+            int count)
+        {
```

Comment:
Yes.. I agree comments would help. This does in-place decryption because the Interop layer is currently scoped to the exact requirements of System.Net.Security.NegotiateStream. The logic of NegotiateStream is to allocate a buffer for Encrypt and do in-place decryption for Decrypt